### ISSUE TITLE
`undocumented_unsafe_blocks` false positive when safety comment is on the previous line

### ISSUE THREAD

**Issue by digama0** [rust-lang/rust-clippy#8449]
### Description

This is somewhat annoying when the project uses minimal-scope `unsafe` blocks. I would prefer for the lint to only trigger if the previous line from the word `unsafe` is not part of a `// Safety` comment. Currently it requires that the comment *directly* precede the `unsafe` token, which leads to a false positive witth very awkward formatting in the suggestion.

```rust
#![warn(clippy::undocumented_unsafe_blocks)]

fn main() {
    let x = &0 as *const _;
    // Safety: x points to a valid value
    println!("{}", unsafe { *x })
}
```
```
warning: unsafe block missing a safety comment
 --> src/main.rs:6:19
  |
6 |     println!("{}", unsafe { *x })
  |                    ^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::undocumented_unsafe_blocks)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#undocumented_unsafe_blocks
help: consider adding a safety comment
  |
6 ~     println!("{}", // SAFETY: ...
7 ~     unsafe { *x })
  |
```

### Version

```text
rustc 1.60.0-nightly (b17226fcc 2022-02-18)
binary: rustc
commit-hash: b17226fcc11587fed612631be372a5b4cb89988a
commit-date: 2022-02-18
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 14.0.0
```

**Comment by digama0** [comment]
If my reading of the code is correct, the reason for this behavior is here:

https://github.com/rust-lang/rust-clippy/blob/29ee5e25d1096f142729ea44c1286e50823762e8/clippy_lints/src/undocumented_unsafe_blocks.rs#L129-L137

it only looks for comments in the enclosing scope. One way to fix this would be to keep going up the syntax tree until you find an enclosing scope that starts on a line different from the unsafe block itself. That would allow things like:
```rust
fn main() {
    // Safety: <- too far away to count
    { // <- this is the enclosing scope within which safety comments are permitted
        // Safety: <- safety comment permitted here 
        let x = { unsafe { foo() } }; // <- surrounding block scope is ignored because it starts on same line
    }
}
```
